#IBM HR Analytics Employee Attrition & Performance
Para essa análise usei as bibliotecas: Pandas , Seaborn, Matplotlib, numpy

In [16]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np


Vou instalar a biblioteca hvplot porque os gráficos são muito bons!

In [41]:
!pip install hvplot
import hvplot.pandas

In [20]:
sns.set_style('darkgrid')
%matplotlib inline
plt.style.use('fivethirtyeight')

pd.set_option('display.float_format', "{:.3f}".format)
pd.set_option('display.max_columns', 80)
pd.set_option('display.max_rows', 80)


Vamos ler o dataset

In [5]:
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

# Análise Exploratória(EDA)

In [8]:
df.shape

(1470, 35)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [23]:
df.describe()

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
count 1470.000   1470.000          1470.000   1470.000       1470.000   
mean    36.924    802.486             9.193      2.913          1.000   
std      9.135    403.509             8.107      1.024          0.000   
min     18.000    102.000             1.000      1.000          1.000   
25%     30.000    465.000             2.000      2.000          1.000   
50%     36.000    802.000             7.000      3.000          1.000   
75%     43.000   1157.000            14.000      4.000          1.000   
max     60.000   1499.000            29.000      5.000          1.000   

       EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
count        1470.000                 1470.000    1470.000        1470.000   
mean         1024.865                    2.722      65.891           2.730   
std           602.024                    1.093      20.329           0.712   
min             1.000                    1.000      30.000           1.000   
25%           491.250                    2.000      48.000           2.000   
50%          1020.500                    3.000      66.000           3.000   
75%          1555.750                    4.000      83.750           3.000   
max          2068.000                    4.000     100.000           4.000   

       JobLevel  JobSatisfaction  MonthlyIncome  MonthlyRate  \
count  1470.000         1470.000       1470.000     1470.000   
mean      2.064            2.729       6502.931    14313.103   
std       1.107            1.103       4707.957     7117.786   
min       1.000            1.000       1009.000     2094.000   
25%       1.000            2.000       2911.000     8047.000   
50%       2.000            3.000       4919.000    14235.500   
75%       3.000            4.000       8379.000    20461.500   
max       5.000            4.000      19999.000    26999.000   

       NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
count            1470.000           1470.000           1470.000   
mean                2.693             15.210              3.154   
std                 2.498              3.660              0.361   
min                 0.000             11.000              3.000   
25%                 1.000             12.000              3.000   
50%                 2.000             14.000              3.000   
75%                 4.000             18.000              3.000   
max                 9.000             25.000              4.000   

       RelationshipSatisfaction  StandardHours  StockOptionLevel  \
count                  1470.000       1470.000          1470.000   
mean                      2.712         80.000             0.794   
std                       1.081          0.000             0.852   
min                       1.000         80.000             0.000   
25%                       2.000         80.000             0.000   
50%                       3.000         80.000             1.000   
75%                       4.000         80.000             1.000   
max                       4.000         80.000             3.000   

       TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
count           1470.000               1470.000         1470.000   
mean              11.280                  2.799            2.761   
std                7.781                  1.289            0.706   
min                0.000                  0.000            1.000   
25%                6.000                  2.000            2.000   
50%               10.000                  3.000            3.000   
75%               15.000                  3.000            3.000   
max               40.000                  6.000            4.000   

       YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
count        1470.000            1470.000                 1470.000   
mean            7.008               4.229                    2.188   
std             6.127            

In [25]:
for column in df.columns:
    print(f"{column}: Number of unique values {df[column].nunique()}")
    print("==================================================")

Age: Number of unique values 43
Attrition: Number of unique values 2
BusinessTravel: Number of unique values 3
DailyRate: Number of unique values 886
Department: Number of unique values 3
DistanceFromHome: Number of unique values 29
Education: Number of unique values 5
EducationField: Number of unique values 6
EmployeeCount: Number of unique values 1
EmployeeNumber: Number of unique values 1470
EnvironmentSatisfaction: Number of unique values 4
Gender: Number of unique values 2
HourlyRate: Number of unique values 71
JobInvolvement: Number of unique values 4
JobLevel: Number of unique values 5
JobRole: Number of unique values 9
JobSatisfaction: Number of unique values 4
MaritalStatus: Number of unique values 3
MonthlyIncome: Number of unique values 1349
MonthlyRate: Number of unique values 1427
NumCompaniesWorked: Number of unique values 10
Over18: Number of unique values 1
OverTime: Number of unique values 2
PercentSalaryHike: Number of unique values 15
PerformanceRating: Number of uni

Percebemos que **'EmployeeCount', 'Over18' e 'StandardHours'** possuem apenas um valor único, e 'EmployeeNumber' possui 1470 valores únicos. Essas características não são úteis para nós, então vamos remover essas colunas.

In [28]:
df.drop(['EmployeeCount', 'Over18', 'StandardHours', 'EmployeeNumber', 'StandardHours'], axis="columns", inplace=True)

Verificação de dados ausentes:

In [5]:
print(df.isnull().sum())

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

# Características Categóricas

In [33]:
object_col = []
for column in df.columns:
    if df[column].dtype == object and len(df[column].unique()) <= 30:
        object_col.append(column)
        print(f"{column} : {df[column].unique()}")
        print(df[column].value_counts())
        print("====================================")
object_col.remove('Attrition')

Attrition : ['Yes' 'No']
Attrition
No     1233
Yes     237
Name: count, dtype: int64
BusinessTravel : ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64
Department : ['Sales' 'Research & Development' 'Human Resources']
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
EducationField : ['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64
Gender : ['Female' 'Male']
Gender
Male      882
Female    588
Name: count, dtype: int64
JobRole : ['Sales Executive' 'Research Scientist' 'Laboratory Technician'
 'Manufacturing Director' 'Healthcare Representative' 'Manager'
 'Sales Representativ

In [34]:
len(object_col)

7

In [38]:
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
df["Attrition"] = label.fit_transform(df.Attrition)

# Características Numéricas

In [39]:
cont_col = []
for column in df.columns:
    if df[column].dtypes != object and df[column].nunique() > 30:
        print(f"{column} : Minimum: {df[column].min()}, Maximum: {df[column].max()}")
        cont_col.append(column)
        print("====================================")

Age : Minimum: 18, Maximum: 60
DailyRate : Minimum: 102, Maximum: 1499
HourlyRate : Minimum: 30, Maximum: 100
MonthlyIncome : Minimum: 1009, Maximum: 19999
MonthlyRate : Minimum: 2094, Maximum: 26999
TotalWorkingYears : Minimum: 0, Maximum: 40
YearsAtCompany : Minimum: 0, Maximum: 40


# Visualização de dados

In [42]:
df.hvplot.hist(y='DistanceFromHome', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [DistanceFromHome]   (Count)

In [43]:
df.hvplot.hist(y='Education', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [Education]   (Count)

In [44]:
df.hvplot.hist(y='RelationshipSatisfaction', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [RelationshipSatisfaction]   (Count)

In [47]:
df.hvplot.hist(y='EnvironmentSatisfaction', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [EnvironmentSatisfaction]   (Count)

In [48]:
df.hvplot.hist(y='JobLevel', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [JobLevel]   (Count)

In [49]:
df.hvplot.hist(y='JobSatisfaction', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [JobSatisfaction]   (Count)

In [50]:
df.hvplot.hist(y='NumCompaniesWorked', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [NumCompaniesWorked]   (Count)

In [51]:
df.hvplot.hist(y='PercentSalaryHike', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [PercentSalaryHike]   (Count)

 *Nota*  
Parece que as variáveis *EnvironmentSatisfaction, JobSatisfaction, PerformanceRating* e *RelationshipSatisfaction* não têm um grande impacto na determinação da *Attrition* dos funcionários.

In [52]:
df.hvplot.hist(y='Age', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [Age]   (Count)

In [53]:
df.hvplot.hist(y='MonthlyIncome', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [MonthlyIncome]   (Count)

In [54]:
df.hvplot.hist(y='YearsAtCompany', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [YearsAtCompany]   (Count)

In [55]:
df.hvplot.hist(y='TotalWorkingYears', by='Attrition', subplots=False, width=600, height=300, bins=30)

:NdOverlay   [Attrition]
   :Histogram   [TotalWorkingYears]   (Count)

#Conclusões:
Os trabalhadores com baixo nível de cargo (JobLevel), renda mensal (MonthlyIncome), anos na empresa (YearsAtCompany) e anos totais de trabalho (TotalWorkingYears) têm maior probabilidade de pedir demissão.

Viagens a trabalho (BusinessTravel): Os funcionários que viajam com frequência têm maior chance de sair do que os outros funcionários.

Departamento (Department): Os funcionários do setor de Pesquisa & Desenvolvimento têm maior probabilidade de permanecer na empresa do que os de outros departamentos.

Área de formação (EducationField): Os trabalhadores com formação em Recursos Humanos e Técnico têm maior probabilidade de sair do que os de outras áreas de educação.

Gênero (Gender): Os homens têm maior probabilidade de pedir demissão.

Cargo (JobRole): Os funcionários nas funções de Técnico de Laboratório, Representante de Vendas e Recursos Humanos têm maior chance de sair do que os de outras posições.

Estado civil (MaritalStatus): Os funcionários solteiros têm maior probabilidade de sair do que os casados e divorciados.

Horas extras (OverTime): Os funcionários que trabalham mais horas têm maior chance de pedir demissão do que os outros.